### PR TITLE
Variadic Generics in asyncio.gather()

### PR DESCRIPTION
**Pull request by Harry-Lees**
`asyncio.gather()` supports an arbitrary number of coroutines as arguments. In versions below 3.11, it wasn't possible to fully type this.

It should now be possible to represent this annotation using Variadic Generics (PEP 646).

### CODE REVIEW

**Comment by AlexWaygood** [comment]
I'm afraid we're a long way off from being able to use PEP 646 in typeshed yet 😞

Pyright supports PEP 646. I think pyre might also, but I'm not sure. Neither mypy nor pytype yet support PEP 646, however, and we would need both to support PEP 646 before we could use variadic generics in typeshed.

File: stdlib/asyncio/tasks.pyi [hunk 1]
```diff
@@ -12,6 +12,7 @@ if sys.version_info >= (3, 9):
     from types import GenericAlias
 if sys.version_info >= (3, 11):
     from contextvars import Context
+    from typing_extensions import TypeVarTuple, Unpack
```
[the remark this answers is not in the document]
Comment (Harry-Lees) [reply stdlib/asyncio/tasks.pyi:15]: > I'm afraid we're a long way off from being able to use PEP 646 in typeshed 😞
> 
> Pyright supports PEP 646. I think pyre might also, but I'm not sure. Neither mypy nor pytype yet support PEP 646, however, and we would need both to support PEP 646 before we could use variadic generics in typeshed.

Yeah, I was actually using Pyright when making this, I noticed on the pipeline that MyPy doesn't support it :(

[the remark this answers is not in the document]
Comment (Harry-Lees) [reply stdlib/asyncio/tasks.pyi:15]: > `TypeVarTuple` and `Unpack` are available from `typing_extensions` on all versions of Python, so (if we could merge this, which we unfortunately can't), you would want to import them from `typing_extensions` unconditionally, and change the signature of `asyncio.gather()` in all Python versions.

Interesting, I didn't actually see that when I was reading the PEP, I actually think that `TypeVarTuple` and `Unpack` should be available in typing from 3.11 onwards? But even Pyright didn't recognise the imports when I tested on 3.11b0

**Comment by Harry-Lees** [comment]
Seeing as this won't be merged, would you like me to close it here, or I can fix the issues you mentioned and just leave it until it's possible to merge?

File: stdlib/asyncio/tasks.pyi (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply stdlib/asyncio/tasks.pyi:15]: > I actually think that `TypeVarTuple` and `Unpack` should be available in typing from 3.11 onwards

Yes, they're available from `typing` in the standard library on 3.11+, and they're available in the third-party backports library `typing_extensions` (which at typeshed, we pretend is part of the stdlib) on 3.6+

**Comment by srittau** [comment]
I think it's fine to prepare this PR. I marked it as "deferred".

**Comment by AlexWaygood** [comment]
> I think it's fine to prepare this PR. I marked it as "deferred".

I'm also fine with this. Be aware, though, that it might be a _long_ time before it gets merged :)

Mypy has https://github.com/python/mypy/issues/12280 to track adding support, but pytype doesn't yet have a tracking issue.

**Comment by AlexWaygood** [comment]
As you can see by the pyright failures, this unfortunately isn't a valid use of PEP 646 as it currently exists. `TypeVarTuple`s, just like normal `TypeVar`s, aren't subscriptable. I think this means that we won't actually be able to use `TypeVarTuple`s for `asyncio.gather()` at all. The type-dependency here doesn't seem to me to be expressable with variadic generics as they currently exist in the type system, due to the fact that we have `tuple[_FutureT[_T1], _FutureT[_T2]]` going in, and `Future[tuple[_FutureT[_T1], _FutureT[_T2]]]` coming out. We have syntax to express `a tuple consisting of a variadic number of TypeVars` (`Unpack[Ts]`), but we don't (AFAIK) have syntax to express `a tuple consisting of a variadic number of objects parameterised by a variadic number of TypeVars`... which I think is what we need here.

(I'm really not an expert on PEP 646 though -- @JelleZijlstra, could you confirm?)

We _can_, I think (once it's supported by mypy/pyright), use PEP 646 to make the signature a little more readable, however. I had a play around, and did a proof-of-concept on my fork over here: https://github.com/AlexWaygood/typeshed/commit/6efe7ec94f6b6f4246f179008e630b6595ae85ea

**Comment by AlexWaygood** [comment]
There's also two other subtleties that are captured by the current signature, but aren't captured by this PR currently:

1. Prior to 3.10, `asyncio.gather` had an optional `loop` parameter.
2. If `return_exceptions` is set to `True`, the return type is different.

**Comment by Harry-Lees** [comment]
> There's also two other subtleties that are captured by the current signature, but aren't captured by this PR currently:
> 
> 1. Prior to 3.10, `asyncio.gather()` had an optional `loop` parameter.
> 2. If `return_exceptions` is set to `True`, the return type is different.

I actually caught these when I removed some of the other `@overload` variants, I didn't immediately fix them because I noticed some other issues, I've just been tinkering in a test file to try and better understand what's going on :)

**Comment by Harry-Lees** [comment]
> We have syntax to express `a tuple consisting of a variadic number of TypeVars` (`Unpack[Ts]`), but we don't (AFAIK) have syntax to express `a tuple consisting of a variadic number of objects parameterised by a variadic number of TypeVars`... which I think is what we need here.

Sorry, I don't quite follow this. Forgive me for explaining back to you, I'm just trying to think it through out loud :P.

You're saying that because we're using the same generic return type `_T` for all the functions in the sequence, we can't represent that each function has a unique return type. For example the following annotation

```python
def example(*args: *_Ts[_T]) -> tuple[*_Ts[_T]]:
```

would be correct for this:

```python
example(1, 2, 3)
```

where the return type would be `tuple[int, int, int]`

but wouldn't be correct for this

```python
example("1", 2, 3, 4, 5)
```

because in the second example the arguments aren't homogeneous?

**Comment by Harry-Lees** [comment]
> due to the fact that we have tuple[_FutureT[_T1], _FutureT[_T2]] going in, and Future[tuple[_FutureT[_T1], _FutureT[_T2]]]

For `gather()` I think we have `tuple[_FutureT[_T1], FutureT[_T2]]` going in and then `Future[tuple[_T1, _T2]]` coming out because `gather()` will just return a list of the results of whatever was passed in.

Also, while we're here

```
Python 3.10.1 (main, Feb  3 2022, 11:22:46) [Clang 12.0.5 (clang-1205.0.22.11)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> async def main():
...     print(await asyncio.gather(asyncio.sleep(1)))
... 
>>> asyncio.run(main())
[None]
```

I think asyncio.gather() returns a list not a tuple, is there a reason that it's annotated as such or should I open a new issue?

**Comment by AlexWaygood** [comment]
> I think asyncio.gather() returns a list not a tuple, is there a reason that it's annotated as such or should I open a new issue?

Yes, this is a deliberate decision due to the fact that the precise type of the parameters is more important here than the precise type of the sequence. See the discussion in #1550.

**Comment by hmc-cs-mdrissi** [comment]
This needs the missing Map operator. There’s a pep draft for type level Map that would handle this but that draft I think is currently on hold and probably wouldn’t mean too much when mypy/pytype first needs to get normal 646.

In a couple years hopefully when we have type level Map then this signature becomes possible. Some builtins like zip and map would also benefit if we had it.

**Comment by AlexWaygood** [comment]
I'd be happy to leave this open if it was just a case of "let's wait for PEP 646 support". But given that it seems like it actually depends on a feature that hasn't even been fully specified yet, let alone had a PEP approved by the Steering Council and been implemented at runtime... I think I'm going to close this for now. But thank you for the PR -- you're absolutely right that the stub is far from perfect at the moment 🙂

**Pull request closed without merging**